pkg/services/annotations/annotationsimpl/loki/alert_state_history_store.go

```diff
@@ -0,0 +1,399 @@
+package loki
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/grafana/grafana/pkg/components/simplejson"
+	"github.com/grafana/grafana/pkg/services/annotations/accesscontrol"
+
+	"github.com/grafana/grafana/pkg/infra/db"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/ngalert/eval"
+	"github.com/grafana/grafana/pkg/services/ngalert/metrics"
+	ngmodels "github.com/grafana/grafana/pkg/services/ngalert/models"
+	"github.com/grafana/grafana/pkg/services/ngalert/state"
+	"github.com/grafana/grafana/pkg/services/ngalert/state/historian"
+	historymodel "github.com/grafana/grafana/pkg/services/ngalert/state/historian/model"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/grafana/grafana/pkg/services/annotations"
+	"github.com/grafana/grafana/pkg/setting"
+	"github.com/grafana/grafana/pkg/util/errutil"
+)
+
+// from grafana/pkg/services/ngalert/state/historian/loki.go
+const defaultQueryRange = 6 * time.Hour
+
+var (
+	ErrLokiStoreInternal    = errutil.NewBase(errutil.StatusInternal, "annotations.loki.internal")
+	ErrLokiStoreUnsupported = errutil.NotImplemented("annotations.loki.unsupported")
+	ErrMissingRule          = errors.New("missing rule")
+)
+
+type lokiQueryClient interface {
+	RangeQuery(ctx context.Context, query string, from, to, limit int64) (historian.QueryRes, error)
+}
+
+type AlertStateHistoryStore struct {
+	client lokiQueryClient
+	db     db.DB
+	log    log.Logger
+}
+
+func NewLokiAlertStateHistoryStore(cfg *setting.Cfg, db db.DB, log log.Logger) (*AlertStateHistoryStore, error) {
```

Comment:
It might be easier for now to just dependency-inject a `LokiConfig`. This system doesn't really care about the wider Grafana settings.

We'll have to probably do this eventually anyway, as there's some fairly complex feature toggle/staging logic in `ngalert.go` that this will need to mesh with. In some Grafana configurations, this component will never be constructed to begin with.

Reply:
This is mimicking the behavior of the xorm store, we're receiving the whole injected config from the parent service. 

My thought was to work on meshing that logic where we construct the store, but if you think there's a better way/place to do that I'm definitely all ears.